### ISSUE TITLE
Produce_data: first impression

### ISSUE THREAD

**Issue by dneise** [cta-sst-1m/digicamtoy#16]
I think digicamtoy.produce_data might very well be the first impression a new user gets from our software chain.

After all it is very difficult to get your hands on real observations. It is also hard to get ones hands on simulations, at least I have not yet found them.

So a natural way for a new user to explore the possibilities of digicampipe might be to create a toy mc file and then use different analyses from digicampipe on that toy mc file. This way a new user can get confidence, that the analyses really work as expected.

I think confidence can make the difference between a succesful software with many users and the rest.

I had a problem executing produce_data.py... it did not work at all.

The problem came from the config file. 
And I think we have to work on that

**Comment by dneise** [comment]
The next thing I did was, studying produce_data.py. I wanted to find out what the config files is really needed for and more importantly I wanted to see how the data is produced.

I found that the data is not really produced inside produce_data.py at all. It seems to be more like a frontend for NTraceGenerator combined with an 'event_sink'.

You know we have these 3 event_sources in digicampipe. One of them can read events from special hdf5 files and produce_data.py writes these hdf5 files. So it implements a special 'event_sink' one could say.

I think this is an important point, since it shows that digicamtoy.produce_data.py and digicampipe.io.hdf5.digicamtoy_event_source form a pair: one cannot modify one without modifying the other.

Therefore I would propose to either merge these 2 repos. Or implement the source and the sink in one package together and people who need the sink import it and use it.

**Comment by dneise** [comment]
So back to the config file. 

I found that two keys play a special role: n_photon and nsb_rate. Both are lists.

In the code of produce_data I found why. Users can produce a lot of different files, with e.g. different nsb rates. They do not have to do multiple calls for this, they just need to specify more values in the nsb_rate list in the config file.

That is a cool feature...

at fiirst glance.

**Comment by dneise** [comment]
At second glance one might ask, what a user wants to vary the electronics noise level? Or the gain?

**Comment by dneise** [comment]
I think a next step should be to move digicamtoy into digicampipe.toy_monte_carlo or so...
certainly a different PR though.

**Comment by dneise** [comment]
However since I canged the sink here ... I also need to change the source in `digicampipe` accrodingly.

It is a small change only, but it needs to be done, if we merge PR #17. the needed change looks like this:
```diff
diff --git a/digicampipe/io/hdf5.py b/digicampipe/io/hdf5.py
index 30cfbf2..95456b1 100644
--- a/digicampipe/io/hdf5.py
+++ b/digicampipe/io/hdf5.py
@@ -25,7 +25,7 @@ def digicamtoy_event_source(
     data = DataContainer()
     hdf5 = h5py.File(url, 'r')
 
-    full_data_set = hdf5['data']['adc_count']
+    full_data_set = hdf5['adc_count']
     n_events, n_pixels, n_samples = full_data_set.shape
 
     if max_events is None:
```

**Comment by calispac** [comment]
I do have the time to look into details #17. But from what you show I seems more simple to produce_data that way. If we want to merge it into master I think for the moment it is better to keep these changes in an other file. 

Actually a feature that is useful in the `NTraceGenerator` is that you can, for almost all parameters, set it per pixel by passing a numpy array (see https://github.com/cta-sst-1m/digicamtoy/blob/master/example.py). Then if you need to configure all parameters pixel-by-pixel it becomes handy to have config files.

**Comment by dneise** [comment]
So after this change, we could write on our README page, that people could do something like this to get started:

    digicam-toy -o test.hdf5
    digicam-view test.hdf5

**Comment by dneise** [comment]
Ah .. I missed one important point. 

"Well ... " , you might say, " .. this is all good and well, but how do we now produce many mc-files with different nsb_rates? Do we really have to type this by hand?"


    $    ./produce_data.py --nsb-rate=0.001 -o nsb_rate_0.001.hdf5
    $    ./produce_data.py --nsb-rate=0.002 -o nsb_rate_0.002.hdf5
    $    ...
    $    ./produce_data.py --nsb-rate=1.000 -o nsb_rate_1.000.hdf5

"But, that's crazy!"

-----

I agree! I think this is an *advanced* monte carlo production task, and therefore needs a bit more advanced methods. What do you think about this?

Fire up your ipython shell (maybe in the folder you'd like to write these monte carlo files into) and type this:
```python
In [1]: from digicamtoy.produce_data import produce_data

In [2]: for nsb_rate in np.logspace(-3, 1, 10):
   ...:     produce_data(
   ...:         nsb_rate=nsb_rate,
   ...:         out_path='nsb_rate{:.3f}.hdf5'.format(nsb_rate)
   ...:    )
```

Would that be okay for you? What do you like better? First producing all 10 config files needed for this and then calling `produce_data.py -y the_correct_config_file.yml`? 
How often did you produce the correct config file, but then somehow made a typo in the call and **called it with the wrong file**. This always happens to me. Writing config files and the calling a program with that config file increases my human error rate a lot.

**Comment by dneise** [comment]
> Actually a feature that is useful in the NTraceGenerator is that you can, for almost all parameters, set it per pixel by passing a numpy array (see https://github.com/cta-sst-1m/digicamtoy/blob/master/example.py). Then if you need to configure all parameters pixel-by-pixel it becomes handy to have config files.

I disagree here, as you can image. Writing these config files is not handy. It is a liability. 
I believe, we and our users become better and better python developers every day. So writing python scripts is easy. If my only way to create the monte carlo files, I want to create,  is by using config files, then this will happen:

  * I write a tool, which creates config files for me.
  * I will wirte a tool which calls produce_data.py with the correct config file for me
  * I will use these tools as a new front end to `produce_data.py`.

If I can simply call a python function, and get my mc-file, I get rid of one step. Instead of writing a tools which creates config files for me, I simply create the configuration (as a dict or so) in mem and give this to that function.

To use the example from above, let's assume somebody really wants to specify different gains and noise levels for all pixels and make a study how this would impact the photon reconstruction for different nsb levels, then maybe they would just write this:
```
In [1]: from digicamtoy.produce_data import produce_data

In [2]: gain = np.random.normal(loc=5.8, scale=0.5, size=1296)

In [3]: sigma_el = np.random.normal(loc=0.8, scale=0.08, size=1296)

In [4]: for nsb_rate in np.logspace(-3, 1, 10):
   ...:     produce_data(
   ...:         nsb_rate=nsb_rate,
   ...:         n_events=10,
   ...:         out_path='indi_gain_and_sigma_el__nsb_rate{:.3f}.hdf5'.format(nsb_rate),
   ...:         gain=gain,
   ...:         sigma_e=sigma_el
   ...:    )
```
